src/com/xilinx/rapidwright/design/DesignTools.java

```diff
@@ -993,6 +992,313 @@ public static boolean removeConnectedRouting(Net net, Node node){
+	            curr = reverseConns.get(sink);
+	            if(curr != null && curr.size() > 1 && atReversedBidirectionalPip) {
+	                for(PIP reversePIP : toRemove) {
+	                    curr.remove(reversePIP);
+	                }
+	            }
+	            atReversedBidirectionalPip = false;
+	            fanoutCount = fanout.get(sink);
+	            fanoutCount = fanoutCount == null ? 0 : fanoutCount;
+	            SitePin sitePin = sink.getSitePin();
+	            if (curr == null && !(sitePin != null || sink.getWireName().contains(Net.VCC_WIRE_NAME) ||
+	                    sink.getWireName().contains(Net.GND_WIRE_NAME))) {
+	                // curr should only be null when we're at the source site, so we've hit a reversed bidirectional PIP
+	                // on our linear path
+	                curr = reverseConnsStart.get(sink);
+	                curr.remove(pip);
+	                fanoutCount--;
+	                atReversedBidirectionalPip = true;
+	            }
+	            if(sitePin != null && sitePin.isInput()){
+	                SiteInst si = p.getSiteInst().getDesign().getSiteInstFromSite(sitePin.getSite());
+	                if(si != null){
+	                    if(net.equals(si.getNetFromSiteWire(sitePin.getPinName()))){
+	                        fanoutCount = 2;
+	                    }
+	                }
+	            }
+	        }
+	        if(curr == null && fanout.size() == 1 && !net.isStaticNet()){
+	            // We got all the way back to the source site. It is likely that 
+	            // the net is using dual exit points from the site as is common in
+	            // SLICEs -- we should unroute the sitenet
+	            SitePin sPin = sink.getSitePin();
+	            SiteInst si = net.getSource().getSiteInst();
+	            BELPin belPin = sPin.getBELPin();
+	            si.unrouteIntraSiteNet(belPin, belPin);
+	        }
+	        p.setRouted(false);	        
+	        for(Node startNode : updateFanout) {
+	            Integer newFanout = fanout.get(startNode);
+	            if(newFanout != null) {
+	                newFanout--;
+	                fanout.put(startNode, newFanout);
+	            }
+	        }
+	    }
+	    ArrayList<PIP> updatedPIPs = new ArrayList<>();
+	    for(PIP pip : net.getPIPs()){
+	        if(!toRemove.contains(pip)) updatedPIPs.add(pip);
+	    }
+	    net.setPIPs(updatedPIPs);
+	}
+
+
+	/**
+	 * This method will completely remove a placed cell (both logical and physical) from a design.
+	 * @param design The design where the cell is instantiated
+	 * @param cell The cell to remove
+	 * @param deferRemovals An optional map that, if passed in non-null will be populated with 
+	 * site pins marked for removal.  The map allows a persistent tracking if this method is called
+	 * many times as the process is expensive without batching. 
+	 */
+	public static void fullyRemoveCell(Design design, Cell cell, Map<Net, Set<SitePinInst>> deferRemovals) {
```

Comment:
Does `deferRemovals` need to be a set? Would you expect a pin to be added more than once since it should only be added when the last cell is removed?

Reply:
> Does `deferRemovals` need to be a set? 
No, only when the caller wants to defer.  

> Would you expect a pin to be added more than once since it should only be added when the last cell is removed?
Pins requiring removal will likely exist for each cell removal.  For example, `SitePinInst` objects will exist for a flop's `Q` output.  In theory we might be able to get away with a List, but in practice, it is probably safer to just use a set as I can't foresee every possible scenario.

Reply:
That's fair. If it's not cut and dry then let's stick with the status quo.